app/models/ticket_purchase.rb

```diff
@@ -25,18 +25,21 @@ class TicketPurchase < ActiveRecord::Base
 
   def self.purchase(conference, user, purchases)
     errors = []
-    ActiveRecord::Base.transaction do
-      conference.tickets.each do |ticket|
-        quantity = purchases[ticket.id.to_s].to_i
-        # if the user bought the ticket and is still unpaid, just update the quantity
-        purchase = if ticket.bought?(user) && ticket.unpaid?(user)
-                     update_quantity(conference, quantity, ticket, user)
-                   else
-                     purchase_ticket(conference, quantity, ticket, user)
-                   end
-
-        if purchase && !purchase.save
-          errors.push(purchase.errors.full_messages)
+    if count_purchased_registration_tickets(conference, purchases) > 1
+      errors.push('You cannot buy more than one registration tickets.')
+    else
+      ActiveRecord::Base.transaction do
+        conference.tickets.each do |ticket|
+          quantity = purchases[ticket.id.to_s].to_i
+          # if the user bought the ticket and is still unpaid, just update the quantity
+          purchase = if ticket.bought?(user) && ticket.unpaid?(user)
+                       update_quantity(conference, quantity, ticket, user)
+                     else
+                       purchase_ticket(conference, quantity, ticket, user)
+                     end
+          if purchase && !purchase.persisted? && !purchase.save
```

Comment:
`!purchase.persisted?` is added because there is a small issue with the way ticket purchase was implemented.
We expect `purchase_ticket` method to return a purchase object with attributes set but the object should not be saved. Well, this works fine for paid tickets but for unpaid tickets we call `pay` method inside `purchase_ticket` method. Inside `pay` method, we use `update_attributes` due to which an already saved purchase object is returned from `purchase_ticket` method. Now this was working fine until now as there is no issue in saving an already saved object. But now with addition of validation `one_registration_ticket_per_user` this code fail in the following case:
**suppose user buys a single registration ticket**, now the first time purchase object is saved inside `pay` method, it will work fine no validation would fail. But the next time the same purchase object is saved here `!purchase.save`, our `one_registration_ticket_per_user` will fail because this time user already has a registration ticket. Though they both represent the same registration ticket.

Reply:
Possible fix:

1. `!purchase.persisted?` so that we don't save an already saved purchase object.

2. Call `pay` method for free ticket somewhere else, outside `purchase_ticket`  method.